File: src/adaos/integrations/inimatic/backend/io/telegram/normalize.ts
import type { ChatInputEvent } from '../types.js'

export function toInputEvent(
	bot_id: string,
	update: any,
	hub_id?: string | null
): ChatInputEvent {
	const upd_id = String(update?.update_id ?? '')

	const hasCb = !!update?.callback_query
	const cb = hasCb ? update.callback_query : undefined
	const msg = update?.message || update?.edited_message || undefined
	const baseMsg = hasCb ? (cb?.message ?? undefined) : msg

	// user_id
	const user_id =
		(baseMsg?.from?.id != null ? String(baseMsg.from.id) : undefined) ??
		(cb?.from?.id != null ? String(cb.from.id) : undefined) ??
		''

	// chat_id
	const chat_id =
		(baseMsg?.chat?.id != null ? String(baseMsg.chat.id) : undefined) ??
		(cb?.message?.chat?.id != null ? String(cb.message.chat.id) : undefined) ??
		''

	// язык (если есть)
	const lang =
		baseMsg?.from?.language_code ??
		cb?.from?.language_code

	// TEXT (/start code и т.п.)
	if (typeof baseMsg?.text === 'string' && baseMsg.text.length > 0) {
		return {
			type: 'text',
			source: 'telegram',
			bot_id,
			hub_id,
			chat_id,
			user_id,
			update_id: upd_id,
			payload: {
				text: baseMsg.text,
				meta: { msg_id: baseMsg.message_id, lang },
			},
		}
	}

	// CALLBACK DATA
	if (typeof cb?.data === 'string' && cb.data.length > 0) {
		return {
			type: 'action',
			source: 'telegram',
			bot_id,
			hub_id,
			chat_id,
			user_id,
			update_id: upd_id,
			payload: { action: { id: cb.data }, meta: { msg_id: cb.message?.message_id } },
		}
	}

	// VOICE
    if (baseMsg?.voice?.file_id) {
        const v = baseMsg.voice
        return {
            type: 'audio',
            source: 'telegram',
            bot_id,
            hub_id,
            chat_id,
            user_id,
            update_id: upd_id,
            payload: {
                file_id: v.file_id,
                // voice messages have no caption in Telegram API; keep meta only
                meta: { msg_id: baseMsg.message_id, mime: 'audio/ogg', duration: v.duration },
            },
        }
    }

	// PHOTO (берём последний размер)
    if (Array.isArray(baseMsg?.photo) && baseMsg.photo.length > 0) {
        const sizes = baseMsg.photo
        const file_id = sizes[sizes.length - 1]?.file_id
        const caption: string | undefined = typeof baseMsg.caption === 'string' && baseMsg.caption.length > 0 ? baseMsg.caption : undefined
        return {
            type: 'photo',
            source: 'telegram',
            bot_id,
            hub_id,
            chat_id,
            user_id,
            update_id: upd_id,
            payload: { file_id, ...(caption ? { text: caption } : {}), meta: { msg_id: baseMsg.message_id } },
        }
    }

	// DOCUMENT
    if (baseMsg?.document?.file_id) {
        const d = baseMsg.document
        const caption: string | undefined = typeof baseMsg.caption === 'string' && baseMsg.caption.length > 0 ? baseMsg.caption : undefined
        return {
            type: 'document',
            source: 'telegram',
            bot_id,
            hub_id,
            chat_id,
            user_id,
            update_id: upd_id,
            payload: { file_id: d.file_id, ...(caption ? { text: caption } : {}), meta: { msg_id: baseMsg.message_id } },
        }
    }

	// UNKNOWN
	return {
		type: 'unknown',
		source: 'telegram',
		bot_id,
		hub_id,
		chat_id,
		user_id,
		update_id: upd_id,
		payload: { meta: { raw_kind: 'unknown', lang } },
	}
}

File: src/adaos/integrations/telegram/normalize.py
from __future__ import annotations
from typing import Mapping, Any, Optional, Dict
from adaos.services.chat_io.interfaces import ChatInputEvent


def to_input_event(bot_id: str, update: Mapping[str, Any], hub_id: Optional[str] = None) -> ChatInputEvent:
    upd_id = str(update.get("update_id", ""))

    # message / edited_message / callback_query
    msg = update.get("message") or update.get("edited_message") or {}
    cb = update.get("callback_query") or {}
    if cb:
        msg = cb.get("message") or {}

    frm = (msg.get("from") or cb.get("from") or {}) or {}
    chat = (msg.get("chat") or {}) or {}
    user_id = str(frm.get("id") or "")
    chat_id = str(chat.get("id") or (cb.get("message") or {}).get("chat", {}).get("id") or "")

    # text
    if msg.get("text"):
        payload: Dict[str, Any] = {
            "text": msg["text"],
            "meta": {"msg_id": msg.get("message_id"), "lang": frm.get("language_code")},
        }
        return ChatInputEvent(
            type="text",
            source="telegram",
            bot_id=bot_id,
            hub_id=hub_id,
            chat_id=chat_id,
            user_id=user_id,
            update_id=upd_id,
            payload=payload,
        )

    # callback_query
    if cb:
        payload = {
            "action": {"id": cb.get("data")},
            "meta": {"msg_id": (cb.get("message") or {}).get("message_id")},
        }
        return ChatInputEvent(
            type="action",
            source="telegram",
            bot_id=bot_id,
            hub_id=hub_id,
            chat_id=chat_id,
            user_id=user_id,
            update_id=upd_id,
            payload=payload,
        )

    # voice (путь заполним после скачивания)
    if msg.get("voice"):
        v = msg["voice"]
        payload = {
            "meta": {"msg_id": msg.get("message_id"), "mime": "audio/ogg", "duration": v.get("duration")},
            "file_id": v.get("file_id"),
        }
        return ChatInputEvent(
            type="audio",
            source="telegram",
            bot_id=bot_id,
            hub_id=hub_id,
            chat_id=chat_id,
            user_id=user_id,
            update_id=upd_id,
            payload=payload,
        )

    # photo
    if msg.get("photo"):
        sizes = msg["photo"]
        file_id = sizes[-1].get("file_id") if sizes else None
        payload = {"file_id": file_id, "meta": {"msg_id": msg.get("message_id")}}
        return ChatInputEvent(
            type="photo",
            source="telegram",
            bot_id=bot_id,
            hub_id=hub_id,
            chat_id=chat_id,
            user_id=user_id,
            update_id=upd_id,
            payload=payload,
        )

    # document
    if msg.get("document"):
        d = msg["document"]
        payload = {"file_id": d.get("file_id"), "meta": {"msg_id": msg.get("message_id")}}
        return ChatInputEvent(
            type="document",
            source="telegram",
            bot_id=bot_id,
            hub_id=hub_id,
            chat_id=chat_id,
            user_id=user_id,
            update_id=upd_id,
            payload=payload,
        )

    # fallback
    return ChatInputEvent(
        type="unknown",
        source="telegram",
        bot_id=bot_id,
        hub_id=hub_id,
        chat_id=chat_id,
        user_id=user_id,
        update_id=upd_id,
        payload={"meta": {"raw_kind": "unknown"}},
    )
